transports/quic/src/connection.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A single QUIC connection.
+//!
+//! The [`Connection`] struct of this module contains, amongst other things, a
+//! [`quinn_proto::Connection`] state machine and an `Arc<Endpoint>`. This struct is responsible
+//! for communication between quinn_proto's connection and its associated endpoint.
+//! All interactions with a QUIC connection should be done through this struct.
+// TODO: docs
+
+use crate::endpoint::{EndpointChannel, ToEndpoint};
+
+use futures::{channel::mpsc, prelude::*};
+use futures_timer::Delay;
+use libp2p_core::PeerId;
+use std::{
+    fmt,
+    net::SocketAddr,
+    task::{Context, Poll},
+    time::Instant,
+};
+
+/// Underlying structure for both [`crate::QuicMuxer`] and [`crate::Upgrade`].
+///
+/// Contains everything needed to process a connection with a remote.
+/// Tied to a specific endpoint.
+pub struct Connection {
+    /// Channel to the endpoint this connection belongs to.
+    endpoint_channel: EndpointChannel,
+    /// Pending message to be sent to the background task that is driving the endpoint.
+    pending_to_endpoint: Option<ToEndpoint>,
+    /// Events that the endpoint will send in destination to our local [`quinn_proto::Connection`].
+    /// Passed at initialization.
+    from_endpoint: mpsc::Receiver<quinn_proto::ConnectionEvent>,
+
+    /// The QUIC state machine for this specific connection.
+    connection: quinn_proto::Connection,
+    /// Identifier for this connection according to the endpoint. Used when sending messages to
+    /// the endpoint.
+    connection_id: quinn_proto::ConnectionHandle,
+    /// `Future` that triggers at the `Instant` that `self.connection.poll_timeout()` indicates.
```

Comment:
```suggestion
    /// `Future` that triggers at the [`Instant`] that `self.connection.poll_timeout()` indicates.
```

In case we ever use something else.